File: frontend/src/api/customer.ts
import apiClient from './client';

// 客户档案管理API接口定义
// 基于客户档案管理模块设计方案v2.md

// 基础数据类型定义
export interface Location {
  type: "Point";
  coordinates: [number, number]; // [longitude, latitude]
}

export interface Meter {
  meter_id: string;
  multiplier: number;
}

export interface MeteringPoint {
  metering_point_id: string;
  allocation_percentage: number; // 0-100
  meter: Meter;
}

export interface UtilityAccount {
  account_id: string;
  metering_points: MeteringPoint[];
}

export interface Customer {
  id: string;
  user_name: string;
  short_name: string;
  user_type?: string;
  industry?: string;
  voltage?: string;
  region?: string;
  district?: string;
  address?: string;
  location?: Location;
  contact_person?: string;
  contact_phone?: string;
  status: 'active' | 'inactive' | 'deleted';
  utility_accounts: UtilityAccount[];
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;
}

export interface CustomerCreate {
  user_name: string;
  short_name: string;
  user_type?: string;
  industry?: string;
  voltage?: string;
  region?: string;
  district?: string;
  address?: string;
  location?: Location;
  contact_person?: string;
  contact_phone?: string;
  utility_accounts?: UtilityAccount[];
}

export interface CustomerUpdate {
  user_name?: string;
  short_name?: string;
  user_type?: string;
  industry?: string;
  voltage?: string;
  region?: string;
  district?: string;
  address?: string;
  location?: Location;
  contact_person?: string;
  contact_phone?: string;
  utility_accounts?: UtilityAccount[];
}

// 查询参数类型定义
// 客户列表项接口（对应后端CustomerListItem）
export interface CustomerListItem {
  id: string;
  user_name: string;
  short_name: string;
  user_type?: string;
  industry?: string;
  voltage?: string;
  region?: string;
  status: 'active' | 'inactive' | 'deleted';
  account_count: number;
  created_at: string;
  updated_at: string;
  // 添加列表中需要的额外字段（虽然在后端CustomerListItem中没有，但前端显示需要）
  contact_person?: string;
  contact_phone?: string;
  // 为了兼容表格中的utility_accounts字段
  utility_accounts?: { length: number };
}

export interface CustomerListParams {
  keyword?: string;
  user_type?: string;
  industry?: string;
  region?: string;
  status?: 'active' | 'inactive' | 'deleted';
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// API接口函数实现

/**
 * 获取客户列表
 * @param params 查询参数
 */
export const getCustomers = (params?: CustomerListParams) => {
  return apiClient.get<PaginatedResponse<CustomerListItem>>('/api/v1/customers', { params });
};

/**
 * 获取客户详情
 * @param customerId 客户ID
 */
export const getCustomer = (customerId: string) => {
  return apiClient.get<Customer>(`/api/v1/customers/${customerId}`);
};

/**
 * 创建新客户
 * @param customerData 客户数据
 */
export const createCustomer = (customerData: CustomerCreate) => {
  return apiClient.post<Customer>('/api/v1/customers', customerData);
};

/**
 * 更新客户信息
 * @param customerId 客户ID
 * @param customerData 更新数据
 */
export const updateCustomer = (customerId: string, customerData: CustomerUpdate) => {
  return apiClient.put<Customer>(`/api/v1/customers/${customerId}`, customerData);
};

/**
 * 删除客户
 * @param customerId 客户ID
 */
export const deleteCustomer = (customerId: string) => {
  return apiClient.delete(`/api/v1/customers/${customerId}`);
};

// 户号管理接口

/**
 * 为客户添加户号
 * @param customerId 客户ID
 * @param accountData 户号数据
 */
export const addCustomerAccount = (customerId: string, accountData: { account_id: string }) => {
  return apiClient.post<UtilityAccount>(`/api/v1/customers/${customerId}/accounts`, accountData);
};

/**
 * 更新客户户号
 * @param customerId 客户ID
 * @param accountId 户号ID
 * @param accountData 更新数据
 */
export const updateCustomerAccount = (
  customerId: string,
  accountId: string,
  accountData: { account_id: string }
) => {
  return apiClient.put<UtilityAccount>(`/api/v1/customers/${customerId}/accounts/${accountId}`, accountData);
};

/**
 * 删除客户户号
 * @param customerId 客户ID
 * @param accountId 户号ID
 */
export const deleteCustomerAccount = (customerId: string, accountId: string) => {
  return apiClient.delete(`/api/v1/customers/${customerId}/accounts/${accountId}`);
};

// 计量点管理接口

/**
 * 为户号添加计量点
 * @param customerId 客户ID
 * @param accountId 户号ID
 * @param meteringPointData 计量点数据
 */
export const addMeteringPoint = (
  customerId: string,
  accountId: string,
  meteringPointData: {
    metering_point_id: string;
    allocation_percentage: number;
    meter: {
      meter_id: string;
      multiplier: number;
    };
  }
) => {
  return apiClient.post<MeteringPoint>(
    `/api/v1/customers/${customerId}/accounts/${accountId}/metering-points`,
    meteringPointData
  );
};

/**
 * 更新计量点信息
 * @param customerId 客户ID
 * @param accountId 户号ID
 * @param meteringPointId 计量点ID
 * @param meteringPointData 更新数据
 */
export const updateMeteringPoint = (
  customerId: string,
  accountId: string,
  meteringPointId: string,
  meteringPointData: {
    metering_point_id?: string;
    allocation_percentage?: number;
    meter?: {
      meter_id?: string;
      multiplier?: number;
    };
  }
) => {
  return apiClient.put<MeteringPoint>(
    `/api/v1/customers/${customerId}/accounts/${accountId}/metering-points/${meteringPointId}`,
    meteringPointData
  );
};

/**
 * 删除计量点
 * @param customerId 客户ID
 * @param accountId 户号ID
 * @param meteringPointId 计量点ID
 */
export const deleteMeteringPoint = (customerId: string, accountId: string, meteringPointId: string) => {
  return apiClient.delete(
    `/api/v1/customers/${customerId}/accounts/${accountId}/metering-points/${meteringPointId}`
  );
};

// 数据一致性管理接口

/**
 * 获取电表信息（用于自动填充）
 * @param meterId 电表ID
 */
export const getMeterInfo = (meterId: string) => {
  return apiClient.get<{
    meter_id: string;
    multiplier: number;
    [key: string]: any;
  }>(`/api/v1/meter-info/${meterId}`);
};

/**
 * 同步更新电表信息
 * @param meterId 电表ID
 * @param updateData 更新数据
 */
export const syncUpdateMeter = (
  meterId: string,
  updateData: {
    multiplier: number;
    sync_all: boolean;
  }
) => {
  return apiClient.post(`/api/v1/meters/${meterId}/sync-update`, updateData);
};

// 常用查询选项
export const USER_TYPES = [
  '标准工商业',
  '大型工商业',
  '农业用户',
  '居民用户'
];

export const INDUSTRIES = [
  '制造业',
  '建筑业',
  '批发和零售业',
  '交通运输业',
  '住宿和餐饮业',
  '信息传输业',
  '金融业',
  '房地产业',
  '租赁和商务服务业',
  '科学研究和技术服务业',
  '水利环境和公共设施管理业',
  '居民服务和其他服务业',
  '教育',
  '卫生和社会工作',
  '文化体育和娱乐业',
  '公共管理和社会组织',
  '国际组织'
];

export const VOLTAGE_LEVELS = [
  '220V',
  '380V',
  '10kV',
  '35kV',
  '110kV',
  '220kV',
  '500kV'
];

export const REGIONS = [
  '南昌市',
  '九江市',
  '景德镇市',
  '萍乡市',
  '新余市',
  '鹰潭市',
  '赣州市',
  '宜春市',
  '上饶市',
  '吉安市',
  '抚州市'
];

export default {
  // 基础CRUD
  getCustomers,
  getCustomer,
  createCustomer,
  updateCustomer,
  deleteCustomer,

  // 户号管理
  addCustomerAccount,
  updateCustomerAccount,
  deleteCustomerAccount,

  // 计量点管理
  addMeteringPoint,
  updateMeteringPoint,
  deleteMeteringPoint,

  // 数据一致性
  getMeterInfo,
  syncUpdateMeter,

  // 常量
  USER_TYPES,
  INDUSTRIES,
  VOLTAGE_LEVELS,
  REGIONS
};

File: webapp/models/customer.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, Literal, List, Any
from datetime import datetime
from bson import ObjectId


class PyObjectId(ObjectId):
    @classmethod
    def __get_pydantic_core_schema__(cls, source_type: Any, handler):
        from pydantic_core import core_schema
        return core_schema.union_schema([
            core_schema.is_instance_schema(ObjectId),
            core_schema.no_info_plain_validator_function(cls.validate),
        ], serialization=core_schema.plain_serializer_function_ser_schema(
            lambda x: str(x)
        ))

    @classmethod
    def validate(cls, v):
        if isinstance(v, ObjectId):
            return v
        if isinstance(v, str):
            if not ObjectId.is_valid(v):
                raise ValueError("Invalid ObjectId")
            return ObjectId(v)
        raise ValueError("Invalid ObjectId")


class BaseMongoModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True
    )

    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")


class Location(BaseModel):
    """地理位置信息"""
    type: Literal["Point"] = "Point"
    coordinates: List[float] = Field(..., description="经纬度坐标 [longitude, latitude]")


class Meter(BaseModel):
    """电表信息"""
    meter_id: str = Field(..., description="电表资产号")
    multiplier: float = Field(..., gt=0, description="倍率")


class MeteringPoint(BaseModel):
    """计量点信息"""
    metering_point_id: str = Field(..., description="计量点ID")
    allocation_percentage: float = Field(..., ge=0, le=100, description="分摊比例(%)")
    meter: Meter = Field(..., description="电表信息")


class UtilityAccount(BaseModel):
    """户号信息"""
    account_id: str = Field(..., description="户号")
    metering_points: List[MeteringPoint] = Field(default_factory=list, description="计量点列表")


class CustomerCreate(BaseModel):
    """客户创建模型"""
    # 客户基本信息
    user_name: str = Field(..., min_length=1, description="客户全称")
    short_name: str = Field(..., min_length=1, description="客户简称")
    user_type: Optional[str] = Field(None, description="客户类型")
    industry: Optional[str] = Field(None, description="行业")
    voltage: Optional[str] = Field(None, description="电压等级")
    region: Optional[str] = Field(None, description="地区")
    district: Optional[str] = Field(None, description="区县")
    address: Optional[str] = Field(None, description="详细地址")
    location: Optional[Location] = Field(None, description="地理位置")

    # 联系信息
    contact_person: Optional[str] = Field(None, description="联系人")
    contact_phone: Optional[str] = Field(None, description="联系电话")

    # 户号信息
    utility_accounts: List[UtilityAccount] = Field(default_factory=list, description="户号列表")


class Customer(BaseMongoModel, CustomerCreate):
    """客户完整模型"""
    # 状态管理（5个核心状态）
    # prospect: 意向客户 - 默认初始状态，未签约
    # pending: 待生效 - 已签约但未开始服务
    # active: 执行中 - 核心活跃状态，正在提供服务
    # suspended: 已暂停 - 临时冻结状态
    # terminated: 已终止 - 合同结束，不可再编辑
    status: Literal["prospect", "pending", "active", "suspended", "terminated"] = "prospect"

    # 审计字段
    created_at: datetime = Field(default_factory=datetime.utcnow, description="创建时间")
    updated_at: datetime = Field(default_factory=datetime.utcnow, description="更新时间")
    created_by: Optional[str] = Field(None, description="创建人")
    updated_by: Optional[str] = Field(None, description="更新人")


class CustomerUpdate(BaseModel):
    """客户更新模型（所有字段可选）"""
    # 客户基本信息
    user_name: Optional[str] = Field(None, min_length=1, description="客户全称")
    short_name: Optional[str] = Field(None, min_length=1, description="客户简称")
    user_type: Optional[str] = Field(None, description="客户类型")
    industry: Optional[str] = Field(None, description="行业")
    voltage: Optional[str] = Field(None, description="电压等级")
    region: Optional[str] = Field(None, description="地区")
    district: Optional[str] = Field(None, description="区县")
    address: Optional[str] = Field(None, description="详细地址")
    location: Optional[Location] = Field(None, description="地理位置")

    # 联系信息
    contact_person: Optional[str] = Field(None, description="联系人")
    contact_phone: Optional[str] = Field(None, description="联系电话")

    # 状态管理
    status: Optional[Literal["prospect", "pending", "active", "suspended", "terminated"]] = Field(None, description="状态")

    # 户号信息
    utility_accounts: Optional[List[UtilityAccount]] = Field(None, description="户号列表")


class CustomerListItem(BaseModel):
    """客户列表项模型"""
    id: str = Field(..., description="客户ID")
    user_name: str = Field(..., description="客户全称")
    user_type: Optional[str] = Field(None, description="客户类型")
    industry: Optional[str] = Field(None, description="行业")
    region: Optional[str] = Field(None, description="地区")
    status: Literal["prospect", "pending", "active", "suspended", "terminated"] = Field(..., description="状态")
    metering_point_count: int = Field(..., description="计量点数量")
    contracted_capacity: Optional[float] = Field(None, description="签约电量(kWh)")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")


class CustomerListResponse(BaseModel):
    """客户列表响应模型"""
    total: int = Field(..., description="总数量")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页大小")
    items: List[CustomerListItem] = Field(..., description="客户列表")


class MeterInfo(BaseModel):
    """电表信息（用于自动填充）"""
    meter_id: str = Field(..., description="电表资产号")
    multiplier: float = Field(..., description="倍率")
    usage_count: int = Field(..., description="使用次数")


class SyncUpdateRequest(BaseModel):
    """同步更新请求模型"""
    multiplier: Optional[float] = Field(None, gt=0, description="新倍率")
    sync_all: bool = Field(True, description="是否同步更新所有计量点")